### ISSUE TITLE
AniDB Cast list problem

### ISSUE THREAD

**Issue by KurtzPT** [ZeroQI/Hama.bundle#225]
Folder: Aikatsu Stars! [anidb-11934]
AniDB Link: https://anidb.net/perl-bin/animedb.pl?show=anime&aid=11934

With AniDB first in the cast preferences, the following actors are downloaded by Hama:

![image](https://user-images.githubusercontent.com/14930057/41944543-2f5b3e7e-79a0-11e8-964e-615a31e19ec3.png)

Notice the 2nd voice actress, Kana. Hama choose her for the voice acting for the character Nikaidou Yuzu.

Looking at AniDB page we have the following:

![image](https://user-images.githubusercontent.com/14930057/41944622-aff3e856-79a0-11e8-9b93-060db6494c03.png)

There's two voice actress for the same character, each one of them has a "role", the first actress role is "Primary Voice", while the second actress role is just a singer for songs.

In this case, Hama should always prioritize the "Primary Voice" over the rest, but it seems it's just taking the last option in the list in each character. This used to be correct before, so something has changed. 

TL;DR: When a character has two voice actors, Hama chooses the last one instead of the "Main Voice Actor"

**Comment by ZeroQI** [comment]
So in the xml: http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=11934
we have:
```
<character id="80808" type="secondary cast in" update="2016-04-07">
<rating votes="1">6.84</rating>
<name>Nikaidou Yuzu</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>185362.jpg</picture>
<seiyuu id="45522" picture="185529.jpg">Kana</seiyuu>
<seiyuu id="33509" picture="204529.jpg">Tadokoro Azusa</seiyuu>
<character id="80808" type="secondary cast in" update="2016-04-07">
<rating votes="1">6.84</rating>
<name>Nikaidou Yuzu</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>185362.jpg</picture>
<seiyuu id="45522" picture="185529.jpg">Kana</seiyuu>
<seiyuu id="33509" picture="204529.jpg">Tadokoro Azusa</seiyuu>
```
It picks the first one but the XML return them in a different order.... 
Code in AniDB section responsible:
```
          for role in xml.xpath('characters/character[(@type="secondary cast in") or (@type="main character in")]'):
          try:
            if GetXml(role, 'seiyuu') and GetXml(role, 'name'):  
              Log.Info('[ ] role: {:<20}, name: {:<20}, photo: {}'.format(GetXml(role, 'name'), GetXml(role, 'seiyuu'), role.get('seiyuu').get('picture') if role.get('seiyuu') else GetXml(role, 'picture') ))
              SaveDict([{'role': role.find('name').text, 'name': role.find('seiyuu').text, 'photo': ANIDB_PIC_BASE_URL + role.find('seiyuu').get('picture')}], AniDB_dict, 'roles')
```
          except Exception as e:  Log.Info("Seyiuu error: {}".format(e))

**Comment by ZeroQI** [comment]
@KurtzPT could you check in multiple AniDB xml if the first AniDB Seyiuu entry is always last in the XML?
- If it is always last in the XML i can change the code to correct this.
- If it is random i am not sure what to do. would like to avoid a list of song voice actors as they may do normal voice acting maybe at times...

**Comment by KurtzPT** [comment]
I'll check it

**Comment by KurtzPT** [comment]
I made a request on AniDB to add that information on their API.

**Comment by ZeroQI** [comment]
@KurtzPT you'e good :). You gave the most precise feedbacks i had in a while to be honest.. If they add that in the API, then it's a piece of cake

**Comment by KurtzPT** [comment]
I know how important good and precise feedback is since I also code (for a living). :)

**Comment by ZeroQI** [comment]
@KurtzPT Well i code for voluntary donations and it doesn't really work...
What do you think of my coding in general? inspired? Am self taught python and tried to make the code reusable and easy. Am most proud of my Dict() function in commons.py

**Comment by KurtzPT** [comment]
I code in other languages, Python isn't something I have experience with, so my knowledge is almost zero of it. (I've been trying to learn it, but lacking time). 

I think you are doing a great job anyway :)

About this issue,I kinda have bad news, I got the following response: " i wouldn't count on it happening soon. not until someone picks up coding for the apis again anyway."

I created the request here: https://tracker.anidb.net/view.php?id=3213 , any news will be on that link I guess.

**Comment by ZeroQI** [comment]
Forum: https://anidb.net/perl-bin/animedb.pl?show=cmt&id=82047
Bug report: https://tracker.anidb.net/view.php?id=3213
- 2018-07-07 10:12 | Ommina | Status | assigned => resolved
- 2018-07-07 10:12 | Ommina | Resolution | open => fixed

HTTP API link: http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=11934

Exemple 1
- Kana (primary: no, singing voice, listed first on web and http API)
- https://anidb.net/perl-bin/animedb.pl?show=character&charid=82392 voiced by: Miyamoto Yume (primary: yes), , 
```
<character id="82392" type="main character in" update="2016-09-22">
<name>Kasumi Mahiru</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>187474.jpg</picture>
<seiyuu id="45522" picture="185529.jpg">Kana</seiyuu>
<seiyuu id="4432" picture="200731.jpg">Miyamoto Yume</seiyuu>
```

Exemple 2:
- voiced by: Asai Ayaka (primary: yes), Rie (primary: no, singing voice, listed second on web and http API)
- https://anidb.net/perl-bin/animedb.pl?show=character&charid=80811 
```
<character id="80811" type="main character in" update="2016-04-07">
<rating votes="1">6.20</rating>
<name>Sakuraba Rola</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>185358.jpg</picture>
<seiyuu id="38232" picture="140073.jpg">Asai Ayaka</seiyuu>
<seiyuu id="45521" picture="185524.jpg">Rie</seiyuu>
```

Here are the two complementary ways to fix it:
- always list primary seiyuu first, no change in fields needed, just a sort, the easiest to implement, i would be delighted with that to fix my bug report 
- add 'primary' and 'comment' fields to improve the HTTP API, the entry would look like: <seiyuu id="4432" picture="185524.jpg" primary ="no" comment="singing voice">Rie</seiyuu>

New AniDB ticket open
https://tracker.anidb.net/view.php?id=3221

**Comment by ZeroQI** [comment]
Closed as duplicated, other issue closed as resolved

**Comment by KurtzPT** [comment]
I saw they changed my ticket to resolved but I didn't noticed anything different in the API response. My guess is that the change is not yet live, but I'm not sure.

**Comment by ZeroQI** [comment]
omina: "I updated the code when I closed the first issue as 'fixed'. It simply isn't running yet."
So at some point the xml should show new field...

**Comment by KurtzPT** [comment]
That explains it 👍

**Comment by ZeroQI** [comment]
When http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=1619 will show new fields for seyiuus in the xml let me know:

```
<characters>
<character id="61486" type="secondary cast in" update="2013-11-14">
<rating votes="3">10.00</rating>
<name>Adberg Erdle</name>
<gender>male</gender>
<charactertype id="1">Character</charactertype>
<picture>145858.jpg</picture>
<seiyuu id="604" picture="154161.jpg">Ogata Ken`ichi</seiyuu>
</character>
```

**Comment by ZeroQI** [comment]
Look like first occurence is the main seiyuu now...
```
<character id="82392" type="main character in" update="2016-09-22">
<name>Kasumi Mahiru</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>187474.jpg</picture>
<seiyuu id="4432" picture="200731.jpg">Miyamoto Yume</seiyuu>
<seiyuu id="45522" picture="185529.jpg">Kana</seiyuu>
</character>
```
and:
```
<character id="80811" type="main character in" update="2016-04-07">
<rating votes="1">6.20</rating>
<name>Sakuraba Rola</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>185358.jpg</picture>
<seiyuu id="38232" picture="140073.jpg">Asai Ayaka</seiyuu>
<seiyuu id="45521" picture="185524.jpg">Rie</seiyuu>
</character>
```
which is what Hama selects...
Can you confirm on series you can tell if wrong if it is correct now?

**Comment by KurtzPT** [comment]
No, just a random coincidence, this one is for example is still at the bottom:

> <character id="80810" type="main character in" update="2016-09-23">
<rating votes="2">9.43</rating>
<name>Saotome Ako</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>187461.jpg</picture>
<seiyuu id="40958" picture="162273.jpg">Miki</seiyuu>
<seiyuu id="45216" picture="184056.jpg">Murakami Natsumi</seiyuu>
</character>

**Comment by ZeroQI** [comment]
AniDB Bug report: https://tracker.anidb.net/view.php?id=3213

http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=11934
- First actress role is "Primary Voice": Murakami Natsumi
```
<character id="80810" type="main character in" update="2016-09-23">
<rating votes="2">9.43</rating>
<name>Saotome Ako</name>
<gender>female</gender>
<charactertype id="1">Character</charactertype>
<picture>187461.jpg</picture>
<seiyuu id="40958" picture="162273.jpg">Miki</seiyuu>
<seiyuu id="45216" picture="184056.jpg">Murakami Natsumi</seiyuu>
</character>
```
other examples reverted to primary voice first

**Comment by KurtzPT** [comment]
I don't think this was fixed at all, despite the ticket being marked as solved. When I wrote the ticket, I suggested an attribute to allow us to understand the role of the seiyuu like for example:
`
<seiyuu id="45216" picture="184056.jpg" role="Primary Voice">Murakami Natsumi</seiyuu>
`

I think the API guy is not active and the modification either wasn't made or it was never put into production. Like this, you can't really make that assumption or any other assumption just based of AnIDB API information.

**Comment by ZeroQI** [comment]
HTTP API show no change...
http://api.anidb.net:9001/httpapi?request=anime&client=hama&clientver=1&protover=1&aid=11934

Closed according to
- https://tracker.anidb.net/view.php?id=3221
- https://tracker.anidb.net/view.php?id=3213

Seem like they changed the order, that's it...
Closing.
